File: android/app/src/test/java/de/sidebyside/next/reference/AttachmentPreparationTest.kt
package de.sidebyside.next.reference

import java.time.OffsetDateTime
import java.util.UUID
import kotlinx.coroutines.test.runTest
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Test
import sidebyside.api.models.AttachmentDetail
import sidebyside.api.models.AttachmentReadRequest
import sidebyside.api.models.AttachmentUploadCreate
import sidebyside.api.models.AuthorSummary
import sidebyside.api.models.MediaType
import sidebyside.api.models.MemoryAttachmentSet
import sidebyside.api.models.MemoryCreate
import sidebyside.api.models.MemoryDetail
import sidebyside.api.models.ReadDescriptor
import sidebyside.api.models.ResourceCapabilities
import sidebyside.api.models.SessionView
import sidebyside.api.models.StoryPage
import sidebyside.api.models.UploadDescriptor

class AttachmentPreparationTest {
    private val spaceId = UUID.fromString("00000000-0000-0000-0000-000000000010")
    private val memoryId = UUID.fromString("00000000-0000-0000-0000-000000000020")
    private val firstAttachmentId = UUID.fromString("00000000-0000-0000-0000-000000000031")
    private val secondAttachmentId = UUID.fromString("00000000-0000-0000-0000-000000000032")
    private val authorId = UUID.fromString("00000000-0000-0000-0000-000000000040")

    @Test
    fun preparesAttachmentBeforeAnyMemoryIsCreated() = runTest {
        val phases = mutableListOf<AttachmentPreparationPhase>()
        var memoryCreated = false
        val api = RecordingContract(
            onCreateMemory = { memoryCreated = true },
            uploadAttachmentId = firstAttachmentId,
        )

        val prepared = prepareAttachment(
            api = api,
            spaceId = spaceId,
            accessToken = "token",
            image = SelectedImage(byteArrayOf(1, 2, 3), "draft.jpg", "image/jpeg"),
            onPhase = phases::add,
        )

        assertFalse(memoryCreated)
        assertEquals(firstAttachmentId, prepared.attachmentId)
        assertEquals(
            listOf(
                AttachmentPreparationPhase.UPLOADING,
                AttachmentPreparationPhase.VALIDATING,
                AttachmentPreparationPhase.READY,
            ),
            phases,
        )
    }

    @Test
    fun bindsPreparedAttachmentsInStableSelectionOrder() = runTest {
        val api = RecordingContract(uploadAttachmentId = firstAttachmentId)

        createMemoryWithPreparedAttachments(
            api = api,
            spaceId = spaceId,
            accessToken = "token",
            title = "Two photos",
            body = "",
            happenedOn = null,
            attachments = listOf(
                PreparedAttachment(firstAttachmentId),
                PreparedAttachment(secondAttachmentId),
            ),
        )

        assertEquals(
            listOf(firstAttachmentId, secondAttachmentId),
            api.boundAttachments!!.attachments.map { it.attachmentId },
        )
        assertEquals(listOf(0, 1), api.boundAttachments!!.attachments.map { it.position })
    }

    private inner class RecordingContract(
        private val onCreateMemory: () -> Unit = {},
        private val uploadAttachmentId: UUID,
    ) : ReferenceContract {
        var boundAttachments: MemoryAttachmentSet? = null

        override suspend fun signIn(email: String, password: String): SessionView = error("not used")

        override suspend fun createMemory(
            spaceId: UUID,
            accessToken: String,
            memory: MemoryCreate,
        ): MemoryDetail {
            onCreateMemory()
            return this@AttachmentPreparationTest.memory()
        }

        override suspend fun createAttachmentUpload(
            spaceId: UUID,
            accessToken: String,
            request: AttachmentUploadCreate,
        ): UploadDescriptor = UploadDescriptor(
            attachment = attachment(uploadAttachmentId, "UPLOADING"),
            method = UploadDescriptor.Method.STREAM,
            requiredHeaders = mapOf("Content-Type" to request.expectedMimeType),
            uploadUrl = "/content",
        )

        override suspend fun uploadAttachmentBytes(
            accessToken: String,
            descriptor: UploadDescriptor,
            image: SelectedImage,
        ) = Unit

        override suspend fun finalizeAttachment(
            spaceId: UUID,
            accessToken: String,
            attachmentId: UUID,
        ): AttachmentDetail = attachment(attachmentId, "VALIDATING")

        override suspend fun getAttachment(
            spaceId: UUID,
            accessToken: String,
            attachmentId: UUID,
        ): AttachmentDetail = attachment(attachmentId, "READY")

        override suspend fun replaceMemoryAttachments(
            spaceId: UUID,
            accessToken: String,
            memoryId: UUID,
            ifMatch: Int,
            attachments: MemoryAttachmentSet,
        ): MemoryDetail {
            boundAttachments = attachments
            return memory(version = 2)
        }

        override suspend fun getTimeline(spaceId: UUID, accessToken: String): StoryPage =
            StoryPage(hasMore = false, items = emptyList(), nextCursor = null)

        override suspend fun createReadAccess(
            spaceId: UUID,
            accessToken: String,
            attachmentId: UUID,
            request: AttachmentReadRequest,
        ): ReadDescriptor = ReadDescriptor(ReadDescriptor.Method.SIGNED_URL, "https://storage.invalid/read")

        override suspend fun readImageBytes(accessToken: String, descriptor: ReadDescriptor): ByteArray =
            byteArrayOf(1, 2, 3)
    }

    private fun attachment(id: UUID, status: String) = AttachmentDetail(
        createdAt = OffsetDateTime.parse("2026-08-29T20:00:00Z"),
        durationSeconds = null,
        hasThumbnail = false,
        height = null,
        id = id,
        mediaType = MediaType.IMAGE,
        mimeType = "image/jpeg",
        propertySize = 3,
        status = status,
        version = 1,
        width = null,
    )

    private fun memory(version: Int = 1) = MemoryDetail(
        attachments = emptyList(),
        author = AuthorSummary(displayName = "A", id = authorId),
        authorId = authorId,
        body = "",
        capabilities = ResourceCapabilities(canComment = true, canDelete = true, canEdit = true),
        createdAt = OffsetDateTime.parse("2026-08-29T20:00:00Z"),
        happenedOn = null,
        id = memoryId,
        spaceId = spaceId,
        title = "Two photos",
        updatedAt = OffsetDateTime.parse("2026-08-29T20:00:00Z"),
        version = version,
    )
}

File: web/src/client/attachmentPreparation.test.ts
import { ReadDescriptorMethodEnum } from '../api/generated/models/ReadDescriptor';
import { UploadDescriptorMethodEnum } from '../api/generated/models/UploadDescriptor';
import {
  createMemoryWithPreparedAttachments,
  prepareAttachment,
  type ReferenceApis,
} from './referenceFlow';

describe('prepareAttachment', () => {
  it('uploads and validates immediately without creating or binding a Memory', async () => {
    const phases: string[] = [];
    const attachment = {
      id: 'attachment-1',
      createdAt: new Date('2026-08-29T20:00:00Z'),
      durationSeconds: null,
      hasThumbnail: false,
      height: null,
      mediaType: 'IMAGE',
      mimeType: 'image/jpeg',
      size: 5,
      status: 'UPLOADING',
      version: 1,
      width: null,
    };
    const createMemory = vi.fn();
    const replaceMemoryAttachments = vi.fn();
    const apis = {
      auth: {},
      memories: { createMemory, replaceMemoryAttachments },
      attachments: {
        createAttachmentUpload: vi.fn(async () => ({
          attachment,
          method: UploadDescriptorMethodEnum.STREAM,
          requiredHeaders: { 'Content-Type': 'image/jpeg' },
          uploadUrl: '/api/v1/spaces/space-1/attachments/attachment-1/content',
        })),
        finalizeAttachmentUpload: vi.fn(async () => ({
          ...attachment,
          status: 'PROCESSING',
        })),
        getAttachment: vi.fn(async () => ({ ...attachment, status: 'READY' })),
      },
      story: {},
    } as unknown as ReferenceApis;
    const fetchApi = vi.fn(async () => new Response(null, { status: 204 })) as unknown as typeof fetch;

    const result = await prepareAttachment(
      apis,
      'https://api.example.invalid',
      'token',
      'space-1',
      new File(['image'], 'test.jpg', { type: 'image/jpeg' }),
      (phase) => phases.push(phase),
      fetchApi,
    );

    expect(phases).toEqual(['uploading', 'validating', 'ready']);
    expect(result).toEqual({ attachmentId: 'attachment-1' });
    expect(createMemory).not.toHaveBeenCalled();
    expect(replaceMemoryAttachments).not.toHaveBeenCalled();
    expect(fetchApi).toHaveBeenCalledOnce();
  });
});

describe('createMemoryWithPreparedAttachments', () => {
  it('binds only the supplied READY attachment identities in stable selection order', async () => {
    const memory = {
      id: 'memory-1',
      version: 4,
      title: 'Two photos',
      body: '',
    };
    const boundMemory = { ...memory, version: 5 };
    const story = { items: [] };
    const apis = {
      auth: {},
      memories: {
        createMemory: vi.fn(async () => memory),
        replaceMemoryAttachments: vi.fn(async () => boundMemory),
      },
      attachments: {
        createAttachmentReadAccess: vi.fn(async () => ({
          method: ReadDescriptorMethodEnum.SIGNED_URL,
          url: 'https://storage.example.invalid/read',
        })),
      },
      story: {
        getStoryTimeline: vi.fn(async () => story),
      },
    } as unknown as ReferenceApis;
    const fetchApi = vi.fn(
      async () =>
        new Response(new Blob(['image'], { type: 'image/jpeg' }), {
          status: 200,
        }),
    ) as unknown as typeof fetch;
    const createObjectUrl = vi
      .spyOn(URL, 'createObjectURL')
      .mockReturnValue('blob:prepared');

    try {
      const result = await createMemoryWithPreparedAttachments(
        apis,
        'https://api.example.invalid',
        'token',
        'space-1',
        { title: memory.title },
        [
          { attachmentId: 'attachment-first' },
          { attachmentId: 'attachment-second' },
        ],
        fetchApi,
      );

      expect(apis.memories.replaceMemoryAttachments).toHaveBeenCalledWith({
        spaceId: 'space-1',
        memoryId: 'memory-1',
        ifMatch: '4',
        memoryAttachmentSet: {
          attachments: [
            { attachmentId: 'attachment-first', position: 0 },
            { attachmentId: 'attachment-second', position: 1 },
          ],
        },
      });
      expect(apis.attachments.createAttachmentReadAccess).toHaveBeenCalledWith({
        spaceId: 'space-1',
        attachmentId: 'attachment-first',
        attachmentReadRequest: {
          parentType: 'MEMORY',
          parentId: 'memory-1',
        },
      });
      expect(result.memory).toBe(boundMemory);
      expect(result.story).toBe(story);
      expect(result.imageUrl).toBe('blob:prepared');
    } finally {
      createObjectUrl.mockRestore();
    }
  });
});
